torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -1163,28 +1163,30 @@ Tensor infinitely_differentiable_logit_backward(
   }
 }
 
-Tensor kl_div_double_backward_grad_output(const Tensor & grad, const Tensor & input, const Tensor & target, int64_t reduction, bool log_target) {
-  auto result = kl_div_backward(grad, input, target, at::Reduction::None, log_target);
+Tensor kl_div_backward(const Tensor& grad, const Tensor& input, const Tensor& target, int64_t reduction, bool log_target) {
+  auto grad_expand = grad.expand_as(input);
+
+  auto grad_input = (log_target ? -at::exp(target) : -target) * grad_expand;
+
   if (reduction == at::Reduction::Mean) {
-    return result.mean();
-  } else if (reduction == at::Reduction::Sum) {
-    return result.sum();
+    grad_input /= input.numel();
   }
-  return result;
+  else if (reduction == at::Reduction::BatchMean) {
+    grad_input /= input.size(0);
+  }
+  return grad_input;
 }
 
-// Compute derivatives for targets.
-Tensor kl_div_target_backward(Tensor grad_output, Tensor self, Tensor target, int64_t reduction, bool log_target) {
-  Tensor grad_target;
-  if (!log_target) {
-    grad_target = grad_output.mul(target.log().add_(1).sub_(self)).masked_fill_(target == 0, 0.);
```

Comment:
The `masked_fill` is actually wrong here. While `x * log(x)` for `x == 0` is commonly defined as `0`, its gradient `1 + log(x)` is `-inf` there.

Reply:
Nit. It's not "commonly defined as 0, but it so happens that lim x -> 0 xlog(x) = 0.